Refactor: Config to Application State

justenwalker opened the pull request:
For #128

This is an initial stab at factoring out the application state from configuration.  One cool thing about this is that many (all?) of the global variables go away.

I also added support for telemetry reload (hopefully, I have not tested this extensively)

I thought I'd get this out in the open so that everyone can poke some holes in it.

cc: @tgross 

## Work in progress

One thing I'm not too happy about is the proliferation of methods that I implemented on the `Config` struct. I think the next step would probably be to push that logic into the packages themselves.  I've had some success implementing configs as some high-level structure containing mostly maps and using [github.com/mitchellh/mapstructure](https://github.com/mitchellh/mapstructure) - since it supports weak typing which would be good for things like #122

tgross commented:
I'm still digging into details as there's a lot here, but just want to say this is some damn nice work @justenwalker

tgross commented:
> One thing I'm not too happy about is the proliferation of methods that I implemented on the Config struct. I think the next step would probably be to push that logic into the packages themselves.

What I've noticed is that there are broadly two categories of methods there. There are some that are thin wrappers on `somepackage.NewSomeStruct` that check for `nil` and that's about it. Those `NewSomeStruct` all take `json.RawMessage`, so maybe there's a nice way to define a function type for those and then push them into the packages as you've suggested.

The second category are those that are building `exec.Cmd` and stuff like that. I suspect we don't want to try to move those out of the `core` package because it's going to cause us heartburn around imports. But moving them into functions that can be tested in isolation (in `core/commands.go` lets say) could be useful.

tgross commented on config/config.go:
```diff
@@ -58,90 +40,19 @@ type Config struct {
 	BackendsConfig  json.RawMessage `json:"backends,omitempty"`
 	TasksConfig     json.RawMessage `json:"tasks,omitempty"`
 	TelemetryConfig json.RawMessage `json:"telemetry,omitempty"`
```
Not necessarily in scope for this PR, but `omitempty` is really only used when we serialize a struct, right? Maybe we don't really need this annotation.

justenwalker replied:
Yeah, we can remove it - we really don't re-serialize the config anyway, so the extra tag options are mostly useless.

tgross commented on config/config.go:
```diff
@@ -5,44 +5,26 @@ import (
+	log "github.com/Sirupsen/logrus"
 )
 
 // Passing around config as a context to functions would be the ideomatic way.
```
I know you're WIP on this, but this whole comment will need reworking.

justenwalker replied:
oops!

tgross commented on core/app.go:
```diff
@@ -0,0 +1,342 @@
+	log.Infof("Reloading configuration.")
+
+	newCfg, err := config.ParseConfig(a.ConfigFlag)
+	if err != nil {
```
We never log this error, which means we fail to reload silently. We should consider doing `log.Errorf` here.

tgross commented on core/app_test.go:
```diff
@@ -229,16 +232,16 @@ func argTestCleanup(oldArgs []string) {
 
 func validateParseError(t *testing.T, input []byte, matchStrings []string) {
-	if cfg, err := unmarshalConfig([]byte(input)); err != nil {
+	if cfg, err := config.UnmarshalConfig([]byte(input)); err != nil {
```
That we had to export this method from the `config` package makes me question whether this test is at the right level of abstraction.

justenwalker replied:
You are correct. I did the minimal work in order to make the tests pass (and some import acrobatics were necessary).
Some of this needs to be re-thought as it is getting hard to test some of this in isolation due to transitive dependencies causing cycles.

tgross commented on core/app.go:
```diff
@@ -0,0 +1,342 @@
+	ConfigFlag       string
+}
+
+// NewApp ...
```
In `Reload()` it looks like we're copying all the content of the newly initialized app on top of the existing `App` struct so that our locks still work? That LGTM as a process but maybe `InitializeApp` and `LoadApp` (and probably `HandlePolling` and `InMaintenanceMode`?) should be non-exported and so that the only public interfaces are:

- `NewApp` (i.e. create from scratch only)
- `Reload`
- `Run`
- `ToggleMaintenanceMode`
- `Terminate`

justenwalker replied:
That sounds good.  I can lock down those other calls

tgross commented on core/signals_test.go:
```diff
@@ -63,42 +62,49 @@ func TestMaintenanceSignal(t *testing.T) {
-	oldConfig.ConfigFlag = `{ "consul": "newconsul:8500" }`
-	if newConfig := reloadConfig(oldConfig); newConfig == nil || newConfig.Consul != "newconsul:8500" {
+	discSvc := app.DiscoveryService
+	if discSvc == nil {
```
Shouldn't we also check for `NoopDiscoveryService`, rather than just that it was cleared to be `nil`?

tgross commented on telemetry/telemetry.go:
```diff
@@ -20,62 +20,77 @@ type Telemetry struct {
+// Serve starts serving the telemetry service
 func (t *Telemetry) Serve() {
-	serverLock.Lock()
-	serverLock.Unlock()
```
oops! That's an embarrassing bug!

justenwalker commented:
@tgross There is a lot here. I broke it out into individual commits that should be hopefully self-explanitory, but I'll summarize:

- I took care of most of the comments that you made in the previous commit
- I did some refactoring of the implementation classes to use `interface{}` instead of `json.RawMessage` for reasons that become a little clearer after you read the source code.
- As a result of using `mapstructure` for decoding configs, I've completed #122 (tested in 9776ebf)
- I put in the ordering fix from #140 (see 2500db4) so that reaping happens first before everything, and signal handling starts before preStart runs.
- Fixed up the `add-dep` target in the makefile, it didn't seem to work when I added `mapstructure`

tgross commented on config/commands.go:
```diff
@@ -0,0 +1,42 @@
+	"os/exec"
+
+	"github.com/joyent/containerpilot/utils"
+	"github.com/prometheus/common/log"
```
prometheus/common/log? What happened to logrus here?

justenwalker replied:
Ah, my auto-importer often gets the common names wrong :(

tgross commented:
One last little comment about the log import in `config/commands.go` but otherwise this LGTM.

tgross commented:
LGTM.
